docs/decisions/0012-semantic-function-multiple-model-support.md

```diff
@@ -0,0 +1,190 @@
+        serviceId: "OpenAIChat", apiKey: oai.ApiKey)
+    .Build();
+
+// Configure semantic function with multiple LLM request settings
+string configPayload = @"{
+  ""schema"": 1,
+  ""description"": ""Hello AI, what can you do for me?"",
+  ""models"": [
+    { ""model_id"": ""text-davinci-003"", ""max_tokens"": 60 },
+    { ""model_id"": ""gpt-35-turbo"", ""max_tokens"": 120 },
+    { ""model_id"": ""gpt-3.5-turbo"", ""max_tokens"": 180 }
+  ]
+}";
+var templateConfig = JsonSerializer.Deserialize<PromptTemplateConfig>(configPayload);
+var func = kernel.CreateSemanticFunction(prompt, config: templateConfig!, "HelloAI");
+
+// Semantic function is executed with OpenAIText using max_tokens=60
+result = await kernel.RunAsync(func);
+```
+
+### Select AI Service and Model Request Settings By Developer Defined Strategy
+
+_As a developer using the Semantic Kernel I can provide factories which select the AI service and request settings used to execute my function so that I can dynamically control which AI service is used to execute my semantic function._
+
+In this case the developer configures different settings based on the model id and provides an AI Service factory which determines which AI Service will be used when the semantic function is executed.
+In the example below the semantic function is executed with whatever AI Service `myServiceFactory` returns.
+
+```csharp
+// Configure a Kernel with multiple LLM's
+IKernel kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithAzureTextCompletionService(deploymentName: aoai.DeploymentName, // text-davinci-003
+        endpoint: aoai.Endpoint, serviceId: "AzureText", apiKey: aoai.ApiKey)
+    .WithAzureChatCompletionService(deploymentName: aoai.ChatDeploymentName, // gpt-35-turbo
+        endpoint: aoai.Endpoint, serviceId: "AzureChat", apiKey: aoai.ApiKey)
+    .WithOpenAITextCompletionService(modelId: oai.ModelId, // text-davinci-003
+        serviceId: "OpenAIText", apiKey: oai.ApiKey, setAsDefault: true)
+    .WithOpenAIChatCompletionService(modelId: oai.ChatModelId, // gpt-3.5-turbo
+        serviceId: "OpenAIChat", apiKey: oai.ApiKey)
+    .Build();
+
+// Configure semantic function with multiple LLM request settings
+string configPayload = @"{
+  ""schema"": 1,
+  ""description"": ""Hello AI, what can you do for me?"",
+  ""models"": [
+    { ""model_id"": ""text-davinci-003"", ""max_tokens"": 60 },
+    { ""model_id"": ""gpt-35-turbo"", ""max_tokens"": 120 },
+    { ""model_id"": ""gpt-3.5-turbo"", ""max_tokens"": 180 }
+  ]
+}";
+var templateConfig = JsonSerializer.Deserialize<PromptTemplateConfig>(configPayload);
+var func = kernel.CreateSemanticFunction(prompt, config: templateConfig!, "HelloAI");
+
+// Semantic function is executed with AI Service returned by custom AIService factory
+var funcOptions = new FunctionOptions();
+funcOptions.AddFactory<IAIService>(myAiServiceFactory);
```

Comment:
I believe the proposed approach of passing an AI service factory as an argument/option of the Kernel.Run async method should be reevaluated and designed properly, with a separate ADR outlining how the design fits/aligns with the existing approach of AI service resolution. For instance, today, it's possible to supply your own implementation of the `IAIServiceProvider` interface when creating the Kernel, so it can easily be used as a factory if needed.

Reply:
Yeah this is just pseudo code to illustrate the concept. I'll remove from the ADR.